bootstrapper/internal/joinclient/joinclient.go

```diff
@@ -209,6 +214,68 @@ func (c *JoinClient) requestJoinTicket(serviceEndpoint string) (ticket *joinprot
 		return nil, nil, err
 	}
 
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		c.log.With(slog.Any("error", err)).Error("Failed to get network interfaces")
+		return nil, nil, err
+	}
+	// Needed since go doesn't implicitly convert slices of structs to slices of interfaces
+	interfacesForFunc := make([]addresses.NetInterface, len(interfaces))
+	for i := range interfaces {
+		interfacesForFunc[i] = &interfaces[i]
+	}
+
+	principalList, err := addresses.GetMachineNetworkAddresses(interfacesForFunc)
+	if err != nil {
+		c.log.With(slog.Any("error", err)).Error("Failed to get network addresses")
+		return nil, nil, err
+	}
+	hostname, err := os.Hostname()
+	if err != nil {
+		c.log.With(slog.Any("error", err)).Error("Failed to get hostname")
+		return nil, nil, err
+	}
+	principalList = append(principalList, hostname)
+
+	var hostKeyPubSSH ssh.PublicKey
+
+	if _, err := c.fileHandler.Stat(constants.SSHHostKeyPath); errors.Is(err, os.ErrNotExist) {
+		hostKeyPub, hostKey, err := ed25519.GenerateKey(nil)
+		if err != nil {
+			c.log.With(slog.Any("error", err)).Error("Failed to generate SSH host key")
+			return nil, nil, err
+		}
+
+		hostKeyPubSSH, err = ssh.NewPublicKey(hostKeyPub)
+		if err != nil {
+			c.log.With(slog.Any("error", err)).Error("Failed to convert ed25519 pubkey to ssh pubkey")
+			return nil, nil, err
+		}
+
+		pemHostKey, err := ssh.MarshalPrivateKey(hostKey, "")
+		if err != nil {
+			c.log.With(slog.Any("error", err)).Error("Failed to format SSH host key")
+			return nil, nil, err
+		}
+		if err := c.fileHandler.Write(constants.SSHHostKeyPath, pem.EncodeToMemory(pemHostKey), file.OptMkdirAll); err != nil {
+			c.log.With(slog.Any("error", err)).Error("Failed to write SSH host key")
+			return nil, nil, err
+		}
+	} else {
+		hostKeyData, err := c.fileHandler.Read(constants.SSHHostKeyPath)
+		if err != nil {
+			c.log.With(slog.Any("error", err)).Error("Failed to read SSH host key file")
+			return nil, nil, err
+		}
+
+		hostKey, err := ssh.ParsePrivateKey(hostKeyData)
+		if err != nil {
+			c.log.With(slog.Any("error", err)).Error("Failed to parse SSH host key file")
+			return nil, nil, err
+		}
+		hostKeyPubSSH = hostKey.PublicKey()
```

Comment:
It looks like this branch assumes an existing `constants.SSHHostKeyPath` file, but 

* `err` might have been something non-nil, other than `os.ErrNotExist`
* the path may be a directory (in principle, obviously unlikely)